File: src/main/java/client/messages/broadcast/AbstractBroadcastMessage.java
package client.messages.broadcast;

import util.packet.PacketWriter;

public abstract class AbstractBroadcastMessage implements BroadcastMessage {

    @Override
    abstract public BroadcastMessageType getType();

    @Override
    public void encode(PacketWriter pw) {
        pw.write(getType().getValue());
        encodeData(pw);
    }

    protected abstract void encodeData(PacketWriter pw);
}

File: src/main/java/client/messages/broadcast/AbstractBroadcastMessage.kt
package client.messages.broadcast

import util.packet.PacketWriter

abstract class AbstractBroadcastMessage : BroadcastMessage {

    abstract override val type: BroadcastMessageType

    override fun encode(pw: PacketWriter) {
        pw.write(type.value)
        encodeData(pw)
    }

    protected abstract fun encodeData(pw: PacketWriter)
}